language/stdlib/modules/FixedPoint32.move

```diff
@@ -44,20 +44,51 @@ module FixedPoint32 {
         assert(product <= MAX_U64, Errors::limit_exceeded(EMULTIPLICATION));
         (product as u64)
     }
+
+    /// We specify the concrete semantics of the implementation but use
+    /// an abstracted, simplified semantics for verification of callers. For the verification outcome of
+    /// callers, the actual result of this function is not relevant, as long as the abstraction behaves
+    /// homomorphic.
     spec fun multiply_u64 {
         pragma opaque;
-        include MultiplyAbortsIf;
-        ensures result == spec_multiply_u64(val, multiplier);
+        include [concrete] ConcreteMultiplyAbortsIf;
+        ensures [concrete] result == spec_concrete_multiply_u64(val, multiplier);
+        include [abstract] MultiplyAbortsIf;
+        ensures [abstract] result == spec_multiply_u64(val, multiplier);
     }
+    spec schema ConcreteMultiplyAbortsIf {
+        val: num;
+        multiplier: FixedPoint32;
+        aborts_if spec_concrete_multiply_u64(val, multiplier) > MAX_U64 with Errors::LIMIT_EXCEEDED;
+    }
+    spec define spec_concrete_multiply_u64(val: num, multiplier: FixedPoint32): num {
+        (val * multiplier.value) >> 32
+    }
+
+    /// # Abstract Semantics
+
     spec schema MultiplyAbortsIf {
         val: num;
         multiplier: FixedPoint32;
         aborts_if spec_multiply_u64(val, multiplier) > MAX_U64 with Errors::LIMIT_EXCEEDED;
     }
+
     spec define spec_multiply_u64(val: num, multiplier: FixedPoint32): num {
-        (val * multiplier.value) >> 32
+        if (multiplier.value == 0)
+            // Zero value
+            0
+        else if (multiplier.value == 1)
+            // 1.0
+            val
+        else if (multiplier.value == 2)
+            // 0.5
+            val / 2
+        else
+            // overflow
+            MAX_U64 + 1
```

Comment:
There can't be any more arithmetic after the overflow case because than it aborts. Indeed as always the aborts is not in the spec function which just is partial, but in the Move function which has the abstracted semantics. So I think this is in fact a sound abstraction in the sense that A(f(x)) == A(f)(f(A(x)), that is if you only observe via the abstracted viewpoint.

Reply:
Well, I think I get your point. Let's say f(x) = x - c. Then the above equation might not hold: A(f(x)) might be defined but A(x) - A(c) is not. It's though interesting to see that this is only module definedness, and I guess there is some kind of theory which captures this kind of refinement module definedness.

Practically, it means that not all computations theoretically possible with Fixedpoint32 are homomorphic for this abstraction. But this would be the case for any kind of abstraction we do perform. For example, we can't prevent someone from doing `fp.value`. We need to lookout that those computations aren't performed in the code we verify.